platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java

```diff
@@ -65,32 +98,85 @@ public void close() throws Exception {
+    try {
+      ConfigurationContextImpl context = createContext(pid, properties);
+      if (context.shouldBeVisibleToPlugins()) {
+        List<ConfigurationPersistencePlugin> plugins =
+            asPlugins(configPersistenceTracker.getServices());
+        for (ConfigurationPersistencePlugin plugin : plugins) {
+          plugin.handleStore(context);
+        }
       }
+      super.store(pid, properties);
+    } finally {
+      serviceStartingLock.unlockRead(stamp);
     }
-    super.store(pid, properties);
   }
 
   @Override
   public void delete(String pid) throws IOException {
-    List<ConfigurationPersistencePlugin> plugins =
-        asPlugins(configPersistenceTracker.getServices());
-    for (ConfigurationPersistencePlugin plugin : plugins) {
-      plugin.handleDelete(pid);
+    // Read lock to ensure plugin initialization is not currently occuring
+    long stamp = serviceStartingLock.readLock();
+    try {
+      List<ConfigurationPersistencePlugin> plugins =
+          asPlugins(configPersistenceTracker.getServices());
+      for (ConfigurationPersistencePlugin plugin : plugins) {
+        plugin.handleDelete(pid);
+      }
+      super.delete(pid);
+    } finally {
+      serviceStartingLock.unlockRead(stamp);
     }
-    super.delete(pid);
   }
 
-  /* Factory method visible for testing purposes only */
+  /* Factory method visible for testing purposes - to inject a mock for the context */
   ConfigurationContextImpl createContext(String pid, Dictionary props) {
     return new ConfigurationContextImpl(pid, props);
   }
 
+  /* Factory method visible for testing purposes - to inject a mock for the plugin */
+  ConfigurationPersistencePlugin retrieveServiceObject(
+      ServiceReference<ConfigurationPersistencePlugin> serviceReference) {
+    return FrameworkUtil.getBundle(DelegatingPersistenceManager.class)
+        .getBundleContext()
+        .getService(serviceReference);
+  }
+
+  private ConfigurationPersistencePlugin handleServiceAdded(
+      ServiceReference<ConfigurationPersistencePlugin> serviceReference) {
+    ConfigurationPersistencePlugin plugin = retrieveServiceObject(serviceReference);
+    // Write lock because no configs should be getting saved or deleted while we initialize
+    long stamp = serviceStartingLock.writeLock();
+    try {
+      Enumeration<Dictionary<String, Object>> dictionaries =
+          getInnerPersistenceManager().getDictionaries();
+      Set<ConfigurationContext> configs = new HashSet<>();
+      while (dictionaries.hasMoreElements()) {
+        Dictionary<String, Object> props = dictionaries.nextElement();
+        ConfigurationContextImpl context =
+            new ConfigurationContextImpl((String) props.get(SERVICE_PID), props);
+        if (context.shouldBeVisibleToPlugins()) {
+          configs.add(context);
+        }
+      }
+      plugin.initialize(configs);
+    } catch (IOException e) {
+      LOGGER.error(
+          format(
```

Comment:
If you are not passing the exception to error(), then use its standard `error(String, Object...)` form instead of calling `format()`. 
It would be nice to trace that stack trace though!

Reply:
Will debug the stack trace. Changed the error log to use it's default interpolation technique.